src/init/chainstate.cpp

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <init/chainstate.h>
+
+#include <consensus/params.h> // for Consensus::Params
+#include <node/blockstorage.h> // for CleanupBlockRevFiles, fHavePruned, fReindex
+#include <validation.h> // for a lot of things
+
+std::optional<ChainstateLoadingError> LoadChainstateSequence(bool fReset,
+                                                             ChainstateManager& chainman,
+                                                             CTxMemPool* mempool,
+                                                             bool fPruneMode,
+                                                             const Consensus::Params& consensus_params,
+                                                             bool fReindexChainState,
+                                                             int64_t nBlockTreeDBCache,
+                                                             int64_t nCoinDBCache,
+                                                             int64_t nCoinCacheUsage,
+                                                             unsigned int check_blocks,
+                                                             unsigned int check_level,
+                                                             bool block_tree_db_in_memory,
+                                                             bool coins_db_in_memory,
+                                                             std::function<int64_t()> get_unix_time_seconds,
+                                                             std::optional<std::function<bool()>> shutdown_requested,
+                                                             std::optional<std::function<void()>> coins_error_cb,
+                                                             std::optional<std::function<void()>> verifying_blocks_cb) {
+    auto is_coinsview_empty = [&](CChainState* chainstate) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
+        return fReset || fReindexChainState || chainstate->CoinsTip().GetBestBlock().IsNull();
+    };
+
+    {
+        LOCK(cs_main);
```

Comment:
This function now consists of three code blocks that each lock `cs_main`. What do you think about just taking `cs_main` at the top of the function and holding it throughout?

Reply:
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6